### PR TITLE
Remove root from iroha init and iroha_client tests.

### PR DESCRIPTION
**Pull request by i1i1**
Moving to genesis block all other codebase.

Signed-off-by: i1i1 <vanyarybin1@live.ru>

### CODE REVIEW

File: iroha/src/genesis.rs [hunk 1]
```diff
@@ -83,7 +82,7 @@ impl GenesisNetwork {
                     .transactions
                     .iter()
                     .map(|raw_transaction| {
-                        raw_transaction.sign_and_accept(&genesis_key_pair, max_instructions_number)
+                        raw_transaction.sign_and_accept(&genesis_key_pair, usize::MAX)
```
Comment (humb1t) [inline iroha/src/genesis.rs:85]: Different machines will put different values here. Are we ok with that?

Comment (i1i1) [reply iroha/src/genesis.rs:85]: I thought that we could make an exception to genesis block, as we trust our self-made json with transactions.

File: iroha/src/queue.rs [hunk 2]
```diff
@@ -75,14 +75,19 @@ impl Queue {
...
             {
                 if let Ok(signature_condition_passed) =
                     transaction.check_signature_condition(world_state_view)
                 {
-                    if is_leader {
-                        if signature_condition_passed {
+                    bad_transaction = false;
+
+                    match (is_leader, signature_condition_passed) {
+                        (false, _) | (true, true) => {
```
Comment (humb1t) [inline iroha/src/queue.rs:90]: Tuples of boolean variables are hard to read and easy to make a mistake.
Please replace with enumeration.

File: iroha/src/torii.rs [hunk 3]
```diff
@@ -181,6 +181,7 @@ async fn handle_queries(
     _query_params: QueryParams,
     request: HttpRequest,
 ) -> Result<HttpResponse, String> {
+    //TODO: Remove when `Result::flatten` https://github.com/rust-lang/rust/issues/70142 will be stabilized
```
Comment (humb1t) [inline iroha/src/torii.rs:184]: Can you also create an issue for that - so we won't forget.

**Review by humb1t — CHANGES_REQUESTED** [review]

File: iroha/src/genesis.rs (hunk 1 above)
Comment (eadventurous) [reply iroha/src/genesis.rs:85]: Though I agree that genesis block is a special case, for consensus simplicity I suggest to try to have the minimal difference for it. So to my mind there also should be a limit for instructions as for any other block.

TL;DR I suggest not to change it here.

File: iroha/src/queue.rs (hunk 2 above)
Comment (i1i1) [reply iroha/src/queue.rs]: Changed to `if !is_leader || signature_condition_passed`

File: iroha/src/init.rs [hunk 4]
```diff
@@ -45,22 +23,37 @@ pub mod config {
...
     #[serde(rename_all = "UPPERCASE")]
     pub struct InitConfiguration {
-        /// The public key of an initial "root@global" account
-        pub root_public_key: PublicKey,
+        /// The user name of initial account
+        pub user_name: String,
+        /// The domain of initial account
+        pub domain_name: String,
+        /// The public key of an initial account
+        pub public_key: PublicKey,
```
Comment (eadventurous) [inline iroha/src/init.rs:38]: Now that we have genesis we don't need any initial accounts apart from the genesis one which is added as part of `GenesisDomain`

File: iroha/src/tx.rs [hunk 5]
```diff
@@ -223,29 +223,42 @@ impl From<RejectedTransaction> for AcceptedTransaction {
...
+        config.genesis_configuration.genesis_block_path = Some(GENESIS_PATH.to_string());
+        config.genesis_configuration.genesis_account_private_key =
+            Some(key_pair.private_key.clone());
+        config.genesis_configuration.genesis_account_public_key = key_pair.public_key.clone();
+
+        let tx = Transaction::new(
+            vec![],
+            AccountId::new(GENESIS_ACCOUNT_NAME, GENESIS_DOMAIN_NAME),
+            1000,
+        );
```
Comment (eadventurous) [inline iroha/src/tx.rs:249]: This test should not be related to Genesis Account at all it is just a test to check transaction hashes. Construct a WSV with some mock data like it is done in queue tests.

File: iroha/src/genesis.rs [hunk 6]
```diff
@@ -242,6 +243,9 @@ pub mod config {
                             format!("Failed to parse Private Key of genesis account: {}", e)
                         })?;
             }
+
+            self.genesis_block_path = env::var(GENESIS_BLOCK_PATH).ok();
+
```
Comment (eadventurous) [inline iroha/src/genesis.rs:248]: Genesis Block should not be loaded from environment its CLI parameter only. It is done so that only the peer that wants to submit the genesis block, will specifically mention it at the start.

File: iroha/src/lib.rs [hunk 7]
```diff
@@ -42,6 +42,9 @@ pub mod tx;
 pub mod world;
 pub mod wsv;
 
+// only for tests
+pub mod peers;
```
Comment (eadventurous) [inline iroha/src/lib.rs:46]: I would suggest to take a look at [idiomatic way to share utility functions discussion](https://stackoverflow.com/questions/44539729/what-is-an-idiomatic-way-to-have-shared-utility-functions-for-integration-tests)

File: iroha/src/lib.rs [hunk 8]
```diff
@@ -210,31 +213,33 @@ impl Iroha {
         let queue = Arc::clone(&self.queue);
         let tx_handle = task::spawn(async move {
             while let Some(transaction) = transactions_receiver.write().await.next().await {
+                log::info!("Pushing tx: {:?}", transaction);
```
Comment (eadventurous) [inline iroha/src/lib.rs:216]: This would drastically increase messages in the console. If it was only used for debugging then remove it, otherwise please decrease log level and also write a more detailed log entry.

File: iroha/src/lib.rs [hunk 9]
```diff
@@ -210,31 +213,33 @@ impl Iroha {
...
-                    task::sleep(TX_RETRIEVAL_INTERVAL).await;
+        let voting_handle = task::spawn(async move {
+            loop {
+                if !sumeragi.write().await.voting_in_progress().await {
+                    let is_leader = sumeragi.read().await.is_leader();
+                    let transactions = queue
+                        .write()
+                        .await
+                        .get_pending_transactions(is_leader, &*world_state_view.read().await);
+
```
Comment (eadventurous) [inline iroha/src/lib.rs:232]: Unnecessary empty lines here.

File: iroha/src/peers.rs [hunk 10]
```diff
@@ -0,0 +1,206 @@
+//! Module for starting peers and networks. Used only for tests
+
```
Comment (eadventurous) [inline iroha/src/peers.rs:2]: I would suggest to call this module/crate test_network

File: iroha/src/peers.rs [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+impl std::cmp::Eq for Peer {}
+
+const CONFIGURATION_PATH: &str = "tests/test_config.json";
+const GENESIS_PATH: &str = "tests/genesis.json";
+
+impl Network {
+    /// Creates new network with some ofline peers
+    pub fn new_offline(
```
Comment (eadventurous) [inline iroha/src/peers.rs:52]:
```suggestion
    pub fn new_with_offline_peers(
```

File: iroha/src/peers.rs [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    pub fn new(
+        default_configuration: Option<Configuration>,
+        n_peers: usize,
+    ) -> Result<Self, String> {
+        Self::new_offline(default_configuration, n_peers, 0)
+    }
+}
+
+impl Peer {
+    fn get_config(&self, configuration: Configuration) -> Configuration {
```
Comment (eadventurous) [inline iroha/src/peers.rs:106]: Please add doc comment.

File: iroha/src/peers.rs [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+    /// Starts peer with config and permissions
+    pub fn start_with_config_permissions(
+        &self,
+        configuration: Configuration,
+        permissions: impl Into<PermissionsValidatorBox> + Send + 'static,
+    ) -> task::JoinHandle<()> {
+        let temp_dir = TempDir::new().expect("Failed to create temp dir");
+        let mut configuration = self.get_config(configuration);
+        let jh = task::spawn(async move {
```
Comment (eadventurous) [inline iroha/src/peers.rs:161]:
```suggestion
        let join_handle = task::spawn(async move {
```

File: iroha/src/sumeragi.rs [hunk 14]
```diff
@@ -158,6 +158,7 @@ impl Sumeragi {
         if transactions.is_empty() {
             return Ok(());
         }
+        log::info!("Rounding txs: {:?}", transactions);
```
Comment (eadventurous) [inline iroha/src/sumeragi.rs:161]: Same thing as with previous log.

File: iroha/src/queue.rs [hunk 15]
```diff
@@ -75,38 +75,31 @@ impl Queue {
...
-                } else {
-                    let _ = self.pending_tx_by_hash.remove(&transaction_hash).expect(
-                        "Failed to get tx by hash. The map should contain txs that are in a queue.",
-                    );
                 }
-            } else {
+            }
+
+            if bad_transaction {
+                log::info!("Removing tx: {:?}", &transaction_hash);
```
Comment (eadventurous) [inline iroha/src/queue.rs:102]: Same thing as with previous log.

File: iroha/src/queue.rs [hunk 16]
```diff
@@ -75,38 +75,31 @@ impl Queue {
...
-                            );
-                            counter -= 1;
-                        }
-                    } else {
-                        output_transactions.push(
-                            self.pending_tx_by_hash
-                                .get(&transaction_hash)
-                                .expect("Failed to get tx by hash. The map should contain txs that are in a queue.")
-                                .clone(),
-                        );
```
Comment (eadventurous) [inline iroha/src/queue.rs:100]: This part seems to break the logic here and decreases readability to my mind better to leave it like it was. I am going to add some comments here to describe what is happening better.

File: iroha_client/tests/account_with_multiple_signatories.rs [hunk 17]
```diff
@@ -1,88 +1,85 @@
-#[cfg(test)]
-mod tests {
```
Comment (eadventurous) [inline iroha_client/tests/account_with_multiple_signatories.rs:2]: Please live mod in all the tests as it was so the diff is better and we can actually understand what changed.

File: iroha_client/tests/add_asset_should_propagate_to_another_peer.rs [hunk 18]
```diff
@@ -1,162 +1,97 @@
-#[cfg(test)]
-mod tests {
```
Comment (eadventurous) [inline iroha_client/tests/add_asset_should_propagate_to_another_peer.rs:2]: Same thing about module. Also do this please for all the other tests.

**Review by eadventurous — CHANGES_REQUESTED** [review]

File: iroha/src/lib.rs (hunk 9 above)
Comment (i1i1) [reply iroha/src/lib.rs:232]: I like to have 1 line separate from variable declaration. Here is just 1 line, so it is ok for `cargo fmt`, but I'll remove it

File: iroha/src/lib.rs (hunk 8 above)
Comment (i1i1) [reply iroha/src/lib.rs:213]: Will remove it as soon as will fix sumeragi tests

File: iroha/src/sumeragi.rs [hunk 19]
```diff
@@ -2002,6 +2009,9 @@ mod tests {
             config.private_key = keys[i].private_key.clone();
             config.public_key = ids[i].public_key.clone();
             config.sumeragi_configuration.trusted_peers(ids.clone());
+            if i == 0 {
+                config.genesis_configuration.genesis_block_path = Some(GENESIS_PATH.to_string());
+            }
```
Comment (eadventurous) [inline iroha/src/sumeragi.rs:2014]: You don't have to submit genesis block in sumeragi tests, just put some mock data into WSV as it is created in these tests.

**Review by eadventurous — CHANGES_REQUESTED** [review]

File: iroha/src/sumeragi.rs [hunk 20]
```diff
@@ -1691,19 +1691,62 @@ mod tests {
...
+        pub const COMMIT_TIME_MS: u64 = 1000;
+        pub const TX_RECEIPT_TIME_MS: u64 = 200;
+        pub const TRANSACTION_TIME_TO_LIVE_MS: u64 = 100_000;
+    }
+
     #[cfg(feature = "network-mock")]
-    const TRANSACTION_TIME_TO_LIVE_MS: u64 = 100_000;
+    fn wsv_create_account(
+        wsv: &mut WorldStateView,
+        genesis_key_pair: &KeyPair,
```
[the remark this answers is not in the document]
Comment (eadventurous) [reply iroha/src/sumeragi.rs:1710]: I think it is just a key_pair for the supplied initial account.

File: iroha/src/sumeragi.rs [hunk 21]
```diff
@@ -2142,10 +2209,19 @@ mod tests {
...
+                if i == 0 {
+                    key_pair = Some(wsv_create_account(
+                        &mut wsv,
+                        &root_key_pair,
+                        account_name,
+                        domain_name,
+                    ));
+                }
+                wsv
+            }));
```
Comment (eadventurous) [inline iroha/src/sumeragi.rs:2224]: Why setup wsv only on leader? All of them should have the same wsv.

Comment (eadventurous) [reply iroha/src/sumeragi.rs:2224]: The same for the tests below.

File: iroha/src/sumeragi.rs [hunk 22]
```diff
@@ -1691,19 +1691,62 @@ mod tests {
...
+                TRANSACTION_TIME_TO_LIVE_MS,
+            )
+            .sign(genesis_key_pair)
+            .expect("Failed to sign."),
+            4096,
+        )
+        .expect("Failed to accept tx.")]);
+
+        let topology = NetworkTopology::new(&wsv.world.trusted_peers_ids, Some(Hash([0u8; 32])), 1)
+            .init()
```
Comment (eadventurous) [inline iroha/src/sumeragi.rs:1739]: It is better not to take a mutable wsv and apply instructions to it, but to `WSV::new(World::new())` and in new `World` setup all the needed data as in queue. Just add there a root@global account and use this wsv for all peers in all tests. No more changes to sumeragi tests are necessary here.

Comment (eadventurous) [reply iroha/src/sumeragi.rs:1739]: Take a look at `queue::tests::world_with_test_domains`

File: scripts/setup_docker_test_env.sh [hunk 23]
```diff
@@ -3,4 +3,4 @@ mkdir test_docker
 cp ./target/debug/iroha_client_cli test_docker
 cp ./iroha_client/config.json test_docker
 docker-compose up -d --build --force-recreate
-sleep 10
+sleep 20
```
Comment (eadventurous) [inline scripts/setup_docker_test_env.sh:6]: Is it necessary?

**Review by humb1t — APPROVED** [review]

File: iroha_client/tests/account_with_multiple_signatories.rs [hunk 24]
```diff
@@ -31,55 +40,49 @@ mod tests {
...
+                        if asset.id.definition_id == asset_definition_id {
+                            return Some(asset);
+                        }
+                    }
+                    None
+                })
+                .next()
+                .expect("Asset should exist.");
+
+            assert_eq!(quantity, asset.quantity);
```
[the remark this answers is not in the document]
Comment (i1i1) [reply iroha_client/tests/account_with_multiple_signatories.rs]: It will break test, as there are multiple assets, and here we are finding asset by id

**Review by eadventurous — APPROVED** [review]
